// typescript/dist/enums/internalSymbolName.enum.js
// Code generated by Herebyfile.mjs generate:enums from internal/ast/symbol.go. DO NOT EDIT.
export var InternalSymbolName;
(function (InternalSymbolName) {
    InternalSymbolName["Call"] = "__call";
    InternalSymbolName["Constructor"] = "__constructor";
    InternalSymbolName["New"] = "__new";
    InternalSymbolName["Index"] = "__index";
    InternalSymbolName["ExportStar"] = "__export";
    InternalSymbolName["Global"] = "__global";
    InternalSymbolName["Missing"] = "__missing";
    InternalSymbolName["Type"] = "__type";
    InternalSymbolName["Object"] = "__object";
    InternalSymbolName["JSXAttributes"] = "__jsxAttributes";
    InternalSymbolName["Class"] = "__class";
    InternalSymbolName["Function"] = "__function";
    InternalSymbolName["Computed"] = "__computed";
    InternalSymbolName["AssignmentDeclaration"] = "__assignment";
    InternalSymbolName["InstantiationExpression"] = "__instantiationExpression";
    InternalSymbolName["ImportAttributes"] = "__importAttributes";
    InternalSymbolName["ExportEquals"] = "export=";
    InternalSymbolName["Default"] = "default";
    InternalSymbolName["This"] = "this";
    InternalSymbolName["ModuleExports"] = "module.exports";
})(InternalSymbolName || (InternalSymbolName = {}));
