// typescript/dist/enums/outerExpressionKinds.js
// Code generated by Herebyfile.mjs generate:enums from internal/ast/utilities.go. DO NOT EDIT.
export var OuterExpressionKinds;
(function (OuterExpressionKinds) {
    OuterExpressionKinds[OuterExpressionKinds["Parentheses"] = 1] = "Parentheses";
    OuterExpressionKinds[OuterExpressionKinds["TypeAssertions"] = 2] = "TypeAssertions";
    OuterExpressionKinds[OuterExpressionKinds["NonNullAssertions"] = 4] = "NonNullAssertions";
    OuterExpressionKinds[OuterExpressionKinds["PartiallyEmittedExpressions"] = 8] = "PartiallyEmittedExpressions";
    OuterExpressionKinds[OuterExpressionKinds["ExpressionsWithTypeArguments"] = 16] = "ExpressionsWithTypeArguments";
    OuterExpressionKinds[OuterExpressionKinds["Satisfies"] = 32] = "Satisfies";
    OuterExpressionKinds[OuterExpressionKinds["ExcludeJSDocTypeAssertion"] = 64] = "ExcludeJSDocTypeAssertion";
    OuterExpressionKinds[OuterExpressionKinds["Assignments"] = 128] = "Assignments";
    OuterExpressionKinds[OuterExpressionKinds["Comma"] = 256] = "Comma";
    OuterExpressionKinds[OuterExpressionKinds["Assertions"] = 38] = "Assertions";
    OuterExpressionKinds[OuterExpressionKinds["All"] = 63] = "All";
    OuterExpressionKinds[OuterExpressionKinds["AllExceptAssertionsOrExpressionsWithTypeArguments"] = 9] = "AllExceptAssertionsOrExpressionsWithTypeArguments";
    OuterExpressionKinds[OuterExpressionKinds["ExpressionTypePassthrough"] = 385] = "ExpressionTypePassthrough";
})(OuterExpressionKinds || (OuterExpressionKinds = {}));
