// typescript/dist/enums/syntaxKind.js
// Code generated by Herebyfile.mjs generate:enums from internal/ast/kind_generated.go. DO NOT EDIT.
export var SyntaxKind;
(function (SyntaxKind) {
    SyntaxKind[SyntaxKind["Unknown"] = 0] = "Unknown";
    SyntaxKind[SyntaxKind["EndOfFile"] = 1] = "EndOfFile";
    SyntaxKind[SyntaxKind["SingleLineCommentTrivia"] = 2] = "SingleLineCommentTrivia";
    SyntaxKind[SyntaxKind["MultiLineCommentTrivia"] = 3] = "MultiLineCommentTrivia";
    SyntaxKind[SyntaxKind["NewLineTrivia"] = 4] = "NewLineTrivia";
    SyntaxKind[SyntaxKind["WhitespaceTrivia"] = 5] = "WhitespaceTrivia";
    SyntaxKind[SyntaxKind["ConflictMarkerTrivia"] = 6] = "ConflictMarkerTrivia";
    SyntaxKind[SyntaxKind["NonTextFileMarkerTrivia"] = 7] = "NonTextFileMarkerTrivia";
    SyntaxKind[SyntaxKind["NumericLiteral"] = 8] = "NumericLiteral";
    SyntaxKind[SyntaxKind["BigIntLiteral"] = 9] = "BigIntLiteral";
    SyntaxKind[SyntaxKind["StringLiteral"] = 10] = "StringLiteral";
    SyntaxKind[SyntaxKind["JsxText"] = 11] = "JsxText";
    SyntaxKind[SyntaxKind["JsxTextAllWhiteSpaces"] = 12] = "JsxTextAllWhiteSpaces";
    SyntaxKind[SyntaxKind["RegularExpressionLiteral"] = 13] = "RegularExpressionLiteral";
    SyntaxKind[SyntaxKind["NoSubstitutionTemplateLiteral"] = 14] = "NoSubstitutionTemplateLiteral";
    SyntaxKind[SyntaxKind["TemplateHead"] = 15] = "TemplateHead";
    SyntaxKind[SyntaxKind["TemplateMiddle"] = 16] = "TemplateMiddle";
    SyntaxKind[SyntaxKind["TemplateTail"] = 17] = "TemplateTail";
    SyntaxKind[SyntaxKind["OpenBraceToken"] = 18] = "OpenBraceToken";
    SyntaxKind[SyntaxKind["CloseBraceToken"] = 19] = "CloseBraceToken";
    SyntaxKind[SyntaxKind["OpenParenToken"] = 20] = "OpenParenToken";
    SyntaxKind[SyntaxKind["CloseParenToken"] = 21] = "CloseParenToken";
    SyntaxKind[SyntaxKind["OpenBracketToken"] = 22] = "OpenBracketToken";
    SyntaxKind[SyntaxKind["CloseBracketToken"] = 23] = "CloseBracketToken";
    SyntaxKind[SyntaxKind["DotToken"] = 24] = "DotToken";
    SyntaxKind[SyntaxKind["DotDotDotToken"] = 25] = "DotDotDotToken";
    SyntaxKind[SyntaxKind["SemicolonToken"] = 26] = "SemicolonToken";
    SyntaxKind[SyntaxKind["CommaToken"] = 27] = "CommaToken";
    SyntaxKind[SyntaxKind["QuestionDotToken"] = 28] = "QuestionDotToken";
    SyntaxKind[SyntaxKind["LessThanToken"] = 29] = "LessThanToken";
    SyntaxKind[SyntaxKind["LessThanSlashToken"] = 30] = "LessThanSlashToken";
    SyntaxKind[SyntaxKind["GreaterThanToken"] = 31] = "GreaterThanToken";
    SyntaxKind[SyntaxKind["LessThanEqualsToken"] = 32] = "LessThanEqualsToken";
    SyntaxKind[SyntaxKind["GreaterThanEqualsToken"] = 33] = "GreaterThanEqualsToken";
    SyntaxKind[SyntaxKind["EqualsEqualsToken"] = 34] = "EqualsEqualsToken";
    SyntaxKind[SyntaxKind["ExclamationEqualsToken"] = 35] = "ExclamationEqualsToken";
    SyntaxKind[SyntaxKind["EqualsEqualsEqualsToken"] = 36] = "EqualsEqualsEqualsToken";
    SyntaxKind[SyntaxKind["ExclamationEqualsEqualsToken"] = 37] = "ExclamationEqualsEqualsToken";
    SyntaxKind[SyntaxKind["EqualsGreaterThanToken"] = 38] = "EqualsGreaterThanToken";
    SyntaxKind[SyntaxKind["PlusToken"] = 39] = "PlusToken";
    SyntaxKind[SyntaxKind["MinusToken"] = 40] = "MinusToken";
    SyntaxKind[SyntaxKind["AsteriskToken"] = 41] = "AsteriskToken";
    SyntaxKind[SyntaxKind["AsteriskAsteriskToken"] = 42] = "AsteriskAsteriskToken";
    SyntaxKind[SyntaxKind["SlashToken"] = 43] = "SlashToken";
    SyntaxKind[SyntaxKind["PercentToken"] = 44] = "PercentToken";
    SyntaxKind[SyntaxKind["PlusPlusToken"] = 45] = "PlusPlusToken";
    SyntaxKind[SyntaxKind["MinusMinusToken"] = 46] = "MinusMinusToken";
    SyntaxKind[SyntaxKind["LessThanLessThanToken"] = 47] = "LessThanLessThanToken";
    SyntaxKind[SyntaxKind["GreaterThanGreaterThanToken"] = 48] = "GreaterThanGreaterThanToken";
    SyntaxKind[SyntaxKind["GreaterThanGreaterThanGreaterThanToken"] = 49] = "GreaterThanGreaterThanGreaterThanToken";
    SyntaxKind[SyntaxKind["AmpersandToken"] = 50] = "AmpersandToken";
    SyntaxKind[SyntaxKind["BarToken"] = 51] = "BarToken";
    SyntaxKind[SyntaxKind["CaretToken"] = 52] = "CaretToken";
    SyntaxKind[SyntaxKind["ExclamationToken"] = 53] = "ExclamationToken";
    SyntaxKind[SyntaxKind["TildeToken"] = 54] = "TildeToken";
    SyntaxKind[SyntaxKind["AmpersandAmpersandToken"] = 55] = "AmpersandAmpersandToken";
    SyntaxKind[SyntaxKind["BarBarToken"] = 56] = "BarBarToken";
    SyntaxKind[SyntaxKind["QuestionToken"] = 57] = "QuestionToken";
    SyntaxKind[SyntaxKind["ColonToken"] = 58] = "ColonToken";
    SyntaxKind[SyntaxKind["AtToken"] = 59] = "AtToken";
    SyntaxKind[SyntaxKind["QuestionQuestionToken"] = 60] = "QuestionQuestionToken";
    SyntaxKind[SyntaxKind["BacktickToken"] = 61] = "BacktickToken";
    SyntaxKind[SyntaxKind["HashToken"] = 62] = "HashToken";
    SyntaxKind[SyntaxKind["EqualsToken"] = 63] = "EqualsToken";
    SyntaxKind[SyntaxKind["PlusEqualsToken"] = 64] = "PlusEqualsToken";
    SyntaxKind[SyntaxKind["MinusEqualsToken"] = 65] = "MinusEqualsToken";
    SyntaxKind[SyntaxKind["AsteriskEqualsToken"] = 66] = "AsteriskEqualsToken";
    SyntaxKind[SyntaxKind["AsteriskAsteriskEqualsToken"] = 67] = "AsteriskAsteriskEqualsToken";
    SyntaxKind[SyntaxKind["SlashEqualsToken"] = 68] = "SlashEqualsToken";
    SyntaxKind[SyntaxKind["PercentEqualsToken"] = 69] = "PercentEqualsToken";
    SyntaxKind[SyntaxKind["LessThanLessThanEqualsToken"] = 70] = "LessThanLessThanEqualsToken";
    SyntaxKind[SyntaxKind["GreaterThanGreaterThanEqualsToken"] = 71] = "GreaterThanGreaterThanEqualsToken";
    SyntaxKind[SyntaxKind["GreaterThanGreaterThanGreaterThanEqualsToken"] = 72] = "GreaterThanGreaterThanGreaterThanEqualsToken";
    SyntaxKind[SyntaxKind["AmpersandEqualsToken"] = 73] = "AmpersandEqualsToken";
    SyntaxKind[SyntaxKind["BarEqualsToken"] = 74] = "BarEqualsToken";
    SyntaxKind[SyntaxKind["BarBarEqualsToken"] = 75] = "BarBarEqualsToken";
    SyntaxKind[SyntaxKind["AmpersandAmpersandEqualsToken"] = 76] = "AmpersandAmpersandEqualsToken";
    SyntaxKind[SyntaxKind["QuestionQuestionEqualsToken"] = 77] = "QuestionQuestionEqualsToken";
    SyntaxKind[SyntaxKind["CaretEqualsToken"] = 78] = "CaretEqualsToken";
    SyntaxKind[SyntaxKind["Identifier"] = 79] = "Identifier";
    SyntaxKind[SyntaxKind["PrivateIdentifier"] = 80] = "PrivateIdentifier";
    SyntaxKind[SyntaxKind["JSDocCommentTextToken"] = 81] = "JSDocCommentTextToken";
    SyntaxKind[SyntaxKind["BreakKeyword"] = 82] = "BreakKeyword";
    SyntaxKind[SyntaxKind["CaseKeyword"] = 83] = "CaseKeyword";
    SyntaxKind[SyntaxKind["CatchKeyword"] = 84] = "CatchKeyword";
    SyntaxKind[SyntaxKind["ClassKeyword"] = 85] = "ClassKeyword";
    SyntaxKind[SyntaxKind["ConstKeyword"] = 86] = "ConstKeyword";
    SyntaxKind[SyntaxKind["ContinueKeyword"] = 87] = "ContinueKeyword";
    SyntaxKind[SyntaxKind["DebuggerKeyword"] = 88] = "DebuggerKeyword";
    SyntaxKind[SyntaxKind["DefaultKeyword"] = 89] = "DefaultKeyword";
    SyntaxKind[SyntaxKind["DeleteKeyword"] = 90] = "DeleteKeyword";
    SyntaxKind[SyntaxKind["DoKeyword"] = 91] = "DoKeyword";
    SyntaxKind[SyntaxKind["ElseKeyword"] = 92] = "ElseKeyword";
    SyntaxKind[SyntaxKind["EnumKeyword"] = 93] = "EnumKeyword";
    SyntaxKind[SyntaxKind["ExportKeyword"] = 94] = "ExportKeyword";
    SyntaxKind[SyntaxKind["ExtendsKeyword"] = 95] = "ExtendsKeyword";
    SyntaxKind[SyntaxKind["FalseKeyword"] = 96] = "FalseKeyword";
    SyntaxKind[SyntaxKind["FinallyKeyword"] = 97] = "FinallyKeyword";
    SyntaxKind[SyntaxKind["ForKeyword"] = 98] = "ForKeyword";
    SyntaxKind[SyntaxKind["FunctionKeyword"] = 99] = "FunctionKeyword";
    SyntaxKind[SyntaxKind["IfKeyword"] = 100] = "IfKeyword";
    SyntaxKind[SyntaxKind["ImportKeyword"] = 101] = "ImportKeyword";
    SyntaxKind[SyntaxKind["InKeyword"] = 102] = "InKeyword";
    SyntaxKind[SyntaxKind["InstanceOfKeyword"] = 103] = "InstanceOfKeyword";
    SyntaxKind[SyntaxKind["NewKeyword"] = 104] = "NewKeyword";
    SyntaxKind[SyntaxKind["NullKeyword"] = 105] = "NullKeyword";
    SyntaxKind[SyntaxKind["ReturnKeyword"] = 106] = "ReturnKeyword";
    SyntaxKind[SyntaxKind["SuperKeyword"] = 107] = "SuperKeyword";
    SyntaxKind[SyntaxKind["SwitchKeyword"] = 108] = "SwitchKeyword";
    SyntaxKind[SyntaxKind["ThisKeyword"] = 109] = "ThisKeyword";
    SyntaxKind[SyntaxKind["ThrowKeyword"] = 110] = "ThrowKeyword";
    SyntaxKind[SyntaxKind["TrueKeyword"] = 111] = "TrueKeyword";
    SyntaxKind[SyntaxKind["TryKeyword"] = 112] = "TryKeyword";
    SyntaxKind[SyntaxKind["TypeOfKeyword"] = 113] = "TypeOfKeyword";
    SyntaxKind[SyntaxKind["VarKeyword"] = 114] = "VarKeyword";
    SyntaxKind[SyntaxKind["VoidKeyword"] = 115] = "VoidKeyword";
    SyntaxKind[SyntaxKind["WhileKeyword"] = 116] = "WhileKeyword";
    SyntaxKind[SyntaxKind["WithKeyword"] = 117] = "WithKeyword";
    SyntaxKind[SyntaxKind["ImplementsKeyword"] = 118] = "ImplementsKeyword";
    SyntaxKind[SyntaxKind["InterfaceKeyword"] = 119] = "InterfaceKeyword";
    SyntaxKind[SyntaxKind["LetKeyword"] = 120] = "LetKeyword";
    SyntaxKind[SyntaxKind["PackageKeyword"] = 121] = "PackageKeyword";
    SyntaxKind[SyntaxKind["PrivateKeyword"] = 122] = "PrivateKeyword";
    SyntaxKind[SyntaxKind["ProtectedKeyword"] = 123] = "ProtectedKeyword";
    SyntaxKind[SyntaxKind["PublicKeyword"] = 124] = "PublicKeyword";
    SyntaxKind[SyntaxKind["StaticKeyword"] = 125] = "StaticKeyword";
    SyntaxKind[SyntaxKind["YieldKeyword"] = 126] = "YieldKeyword";
    SyntaxKind[SyntaxKind["AbstractKeyword"] = 127] = "AbstractKeyword";
    SyntaxKind[SyntaxKind["AccessorKeyword"] = 128] = "AccessorKeyword";
    SyntaxKind[SyntaxKind["AsKeyword"] = 129] = "AsKeyword";
    SyntaxKind[SyntaxKind["AssertsKeyword"] = 130] = "AssertsKeyword";
    SyntaxKind[SyntaxKind["AssertKeyword"] = 131] = "AssertKeyword";
    SyntaxKind[SyntaxKind["AnyKeyword"] = 132] = "AnyKeyword";
    SyntaxKind[SyntaxKind["AsyncKeyword"] = 133] = "AsyncKeyword";
    SyntaxKind[SyntaxKind["AwaitKeyword"] = 134] = "AwaitKeyword";
    SyntaxKind[SyntaxKind["BooleanKeyword"] = 135] = "BooleanKeyword";
    SyntaxKind[SyntaxKind["ConstructorKeyword"] = 136] = "ConstructorKeyword";
    SyntaxKind[SyntaxKind["DeclareKeyword"] = 137] = "DeclareKeyword";
    SyntaxKind[SyntaxKind["GetKeyword"] = 138] = "GetKeyword";
    SyntaxKind[SyntaxKind["ImmediateKeyword"] = 139] = "ImmediateKeyword";
    SyntaxKind[SyntaxKind["InferKeyword"] = 140] = "InferKeyword";
    SyntaxKind[SyntaxKind["IntrinsicKeyword"] = 141] = "IntrinsicKeyword";
    SyntaxKind[SyntaxKind["IsKeyword"] = 142] = "IsKeyword";
    SyntaxKind[SyntaxKind["KeyOfKeyword"] = 143] = "KeyOfKeyword";
    SyntaxKind[SyntaxKind["ModuleKeyword"] = 144] = "ModuleKeyword";
    SyntaxKind[SyntaxKind["NamespaceKeyword"] = 145] = "NamespaceKeyword";
    SyntaxKind[SyntaxKind["NeverKeyword"] = 146] = "NeverKeyword";
    SyntaxKind[SyntaxKind["OutKeyword"] = 147] = "OutKeyword";
    SyntaxKind[SyntaxKind["ReadonlyKeyword"] = 148] = "ReadonlyKeyword";
    SyntaxKind[SyntaxKind["RequireKeyword"] = 149] = "RequireKeyword";
    SyntaxKind[SyntaxKind["NumberKeyword"] = 150] = "NumberKeyword";
    SyntaxKind[SyntaxKind["ObjectKeyword"] = 151] = "ObjectKeyword";
    SyntaxKind[SyntaxKind["SatisfiesKeyword"] = 152] = "SatisfiesKeyword";
    SyntaxKind[SyntaxKind["SetKeyword"] = 153] = "SetKeyword";
    SyntaxKind[SyntaxKind["StringKeyword"] = 154] = "StringKeyword";
    SyntaxKind[SyntaxKind["SymbolKeyword"] = 155] = "SymbolKeyword";
    SyntaxKind[SyntaxKind["TypeKeyword"] = 156] = "TypeKeyword";
    SyntaxKind[SyntaxKind["UndefinedKeyword"] = 157] = "UndefinedKeyword";
    SyntaxKind[SyntaxKind["UniqueKeyword"] = 158] = "UniqueKeyword";
    SyntaxKind[SyntaxKind["UnknownKeyword"] = 159] = "UnknownKeyword";
    SyntaxKind[SyntaxKind["UsingKeyword"] = 160] = "UsingKeyword";
    SyntaxKind[SyntaxKind["FromKeyword"] = 161] = "FromKeyword";
    SyntaxKind[SyntaxKind["GlobalKeyword"] = 162] = "GlobalKeyword";
    SyntaxKind[SyntaxKind["BigIntKeyword"] = 163] = "BigIntKeyword";
    SyntaxKind[SyntaxKind["OverrideKeyword"] = 164] = "OverrideKeyword";
    SyntaxKind[SyntaxKind["OfKeyword"] = 165] = "OfKeyword";
    SyntaxKind[SyntaxKind["DeferKeyword"] = 166] = "DeferKeyword";
    SyntaxKind[SyntaxKind["QualifiedName"] = 167] = "QualifiedName";
    SyntaxKind[SyntaxKind["ComputedPropertyName"] = 168] = "ComputedPropertyName";
    SyntaxKind[SyntaxKind["TypeParameter"] = 169] = "TypeParameter";
    SyntaxKind[SyntaxKind["Parameter"] = 170] = "Parameter";
    SyntaxKind[SyntaxKind["Decorator"] = 171] = "Decorator";
    SyntaxKind[SyntaxKind["PropertySignature"] = 172] = "PropertySignature";
    SyntaxKind[SyntaxKind["PropertyDeclaration"] = 173] = "PropertyDeclaration";
    SyntaxKind[SyntaxKind["MethodSignature"] = 174] = "MethodSignature";
    SyntaxKind[SyntaxKind["MethodDeclaration"] = 175] = "MethodDeclaration";
    SyntaxKind[SyntaxKind["ClassStaticBlockDeclaration"] = 176] = "ClassStaticBlockDeclaration";
    SyntaxKind[SyntaxKind["Constructor"] = 177] = "Constructor";
    SyntaxKind[SyntaxKind["GetAccessor"] = 178] = "GetAccessor";
    SyntaxKind[SyntaxKind["SetAccessor"] = 179] = "SetAccessor";
    SyntaxKind[SyntaxKind["CallSignature"] = 180] = "CallSignature";
    SyntaxKind[SyntaxKind["ConstructSignature"] = 181] = "ConstructSignature";
    SyntaxKind[SyntaxKind["IndexSignature"] = 182] = "IndexSignature";
    SyntaxKind[SyntaxKind["TypePredicate"] = 183] = "TypePredicate";
    SyntaxKind[SyntaxKind["TypeReference"] = 184] = "TypeReference";
    SyntaxKind[SyntaxKind["FunctionType"] = 185] = "FunctionType";
    SyntaxKind[SyntaxKind["ConstructorType"] = 186] = "ConstructorType";
    SyntaxKind[SyntaxKind["TypeQuery"] = 187] = "TypeQuery";
    SyntaxKind[SyntaxKind["TypeLiteral"] = 188] = "TypeLiteral";
    SyntaxKind[SyntaxKind["ArrayType"] = 189] = "ArrayType";
    SyntaxKind[SyntaxKind["TupleType"] = 190] = "TupleType";
    SyntaxKind[SyntaxKind["OptionalType"] = 191] = "OptionalType";
    SyntaxKind[SyntaxKind["RestType"] = 192] = "RestType";
    SyntaxKind[SyntaxKind["UnionType"] = 193] = "UnionType";
    SyntaxKind[SyntaxKind["IntersectionType"] = 194] = "IntersectionType";
    SyntaxKind[SyntaxKind["ConditionalType"] = 195] = "ConditionalType";
    SyntaxKind[SyntaxKind["InferType"] = 196] = "InferType";
    SyntaxKind[SyntaxKind["ParenthesizedType"] = 197] = "ParenthesizedType";
    SyntaxKind[SyntaxKind["ThisType"] = 198] = "ThisType";
    SyntaxKind[SyntaxKind["TypeOperator"] = 199] = "TypeOperator";
    SyntaxKind[SyntaxKind["IndexedAccessType"] = 200] = "IndexedAccessType";
    SyntaxKind[SyntaxKind["MappedType"] = 201] = "MappedType";
    SyntaxKind[SyntaxKind["LiteralType"] = 202] = "LiteralType";
    SyntaxKind[SyntaxKind["NamedTupleMember"] = 203] = "NamedTupleMember";
    SyntaxKind[SyntaxKind["TemplateLiteralType"] = 204] = "TemplateLiteralType";
    SyntaxKind[SyntaxKind["TemplateLiteralTypeSpan"] = 205] = "TemplateLiteralTypeSpan";
    SyntaxKind[SyntaxKind["ImportType"] = 206] = "ImportType";
    SyntaxKind[SyntaxKind["ObjectBindingPattern"] = 207] = "ObjectBindingPattern";
    SyntaxKind[SyntaxKind["ArrayBindingPattern"] = 208] = "ArrayBindingPattern";
    SyntaxKind[SyntaxKind["BindingElement"] = 209] = "BindingElement";
    SyntaxKind[SyntaxKind["ArrayLiteralExpression"] = 210] = "ArrayLiteralExpression";
    SyntaxKind[SyntaxKind["ObjectLiteralExpression"] = 211] = "ObjectLiteralExpression";
    SyntaxKind[SyntaxKind["PropertyAccessExpression"] = 212] = "PropertyAccessExpression";
    SyntaxKind[SyntaxKind["ElementAccessExpression"] = 213] = "ElementAccessExpression";
    SyntaxKind[SyntaxKind["CallExpression"] = 214] = "CallExpression";
    SyntaxKind[SyntaxKind["NewExpression"] = 215] = "NewExpression";
    SyntaxKind[SyntaxKind["TaggedTemplateExpression"] = 216] = "TaggedTemplateExpression";
    SyntaxKind[SyntaxKind["TypeAssertionExpression"] = 217] = "TypeAssertionExpression";
    SyntaxKind[SyntaxKind["ParenthesizedExpression"] = 218] = "ParenthesizedExpression";
    SyntaxKind[SyntaxKind["FunctionExpression"] = 219] = "FunctionExpression";
    SyntaxKind[SyntaxKind["ArrowFunction"] = 220] = "ArrowFunction";
    SyntaxKind[SyntaxKind["DeleteExpression"] = 221] = "DeleteExpression";
    SyntaxKind[SyntaxKind["TypeOfExpression"] = 222] = "TypeOfExpression";
    SyntaxKind[SyntaxKind["VoidExpression"] = 223] = "VoidExpression";
    SyntaxKind[SyntaxKind["AwaitExpression"] = 224] = "AwaitExpression";
    SyntaxKind[SyntaxKind["PrefixUnaryExpression"] = 225] = "PrefixUnaryExpression";
    SyntaxKind[SyntaxKind["PostfixUnaryExpression"] = 226] = "PostfixUnaryExpression";
    SyntaxKind[SyntaxKind["BinaryExpression"] = 227] = "BinaryExpression";
    SyntaxKind[SyntaxKind["ConditionalExpression"] = 228] = "ConditionalExpression";
    SyntaxKind[SyntaxKind["TemplateExpression"] = 229] = "TemplateExpression";
    SyntaxKind[SyntaxKind["YieldExpression"] = 230] = "YieldExpression";
    SyntaxKind[SyntaxKind["SpreadElement"] = 231] = "SpreadElement";
    SyntaxKind[SyntaxKind["ClassExpression"] = 232] = "ClassExpression";
    SyntaxKind[SyntaxKind["OmittedExpression"] = 233] = "OmittedExpression";
    SyntaxKind[SyntaxKind["ExpressionWithTypeArguments"] = 234] = "ExpressionWithTypeArguments";
    SyntaxKind[SyntaxKind["AsExpression"] = 235] = "AsExpression";
    SyntaxKind[SyntaxKind["NonNullExpression"] = 236] = "NonNullExpression";
    SyntaxKind[SyntaxKind["MetaProperty"] = 237] = "MetaProperty";
    SyntaxKind[SyntaxKind["SyntheticExpression"] = 238] = "SyntheticExpression";
    SyntaxKind[SyntaxKind["SatisfiesExpression"] = 239] = "SatisfiesExpression";
    SyntaxKind[SyntaxKind["TemplateSpan"] = 240] = "TemplateSpan";
    SyntaxKind[SyntaxKind["SemicolonClassElement"] = 241] = "SemicolonClassElement";
    SyntaxKind[SyntaxKind["Block"] = 242] = "Block";
    SyntaxKind[SyntaxKind["EmptyStatement"] = 243] = "EmptyStatement";
    SyntaxKind[SyntaxKind["VariableStatement"] = 244] = "VariableStatement";
    SyntaxKind[SyntaxKind["ExpressionStatement"] = 245] = "ExpressionStatement";
    SyntaxKind[SyntaxKind["IfStatement"] = 246] = "IfStatement";
    SyntaxKind[SyntaxKind["DoStatement"] = 247] = "DoStatement";
    SyntaxKind[SyntaxKind["WhileStatement"] = 248] = "WhileStatement";
    SyntaxKind[SyntaxKind["ForStatement"] = 249] = "ForStatement";
    SyntaxKind[SyntaxKind["ForInStatement"] = 250] = "ForInStatement";
    SyntaxKind[SyntaxKind["ForOfStatement"] = 251] = "ForOfStatement";
    SyntaxKind[SyntaxKind["ContinueStatement"] = 252] = "ContinueStatement";
    SyntaxKind[SyntaxKind["BreakStatement"] = 253] = "BreakStatement";
    SyntaxKind[SyntaxKind["ReturnStatement"] = 254] = "ReturnStatement";
    SyntaxKind[SyntaxKind["WithStatement"] = 255] = "WithStatement";
    SyntaxKind[SyntaxKind["SwitchStatement"] = 256] = "SwitchStatement";
    SyntaxKind[SyntaxKind["LabeledStatement"] = 257] = "LabeledStatement";
    SyntaxKind[SyntaxKind["ThrowStatement"] = 258] = "ThrowStatement";
    SyntaxKind[SyntaxKind["TryStatement"] = 259] = "TryStatement";
    SyntaxKind[SyntaxKind["DebuggerStatement"] = 260] = "DebuggerStatement";
    SyntaxKind[SyntaxKind["VariableDeclaration"] = 261] = "VariableDeclaration";
    SyntaxKind[SyntaxKind["VariableDeclarationList"] = 262] = "VariableDeclarationList";
    SyntaxKind[SyntaxKind["FunctionDeclaration"] = 263] = "FunctionDeclaration";
    SyntaxKind[SyntaxKind["ClassDeclaration"] = 264] = "ClassDeclaration";
    SyntaxKind[SyntaxKind["InterfaceDeclaration"] = 265] = "InterfaceDeclaration";
    SyntaxKind[SyntaxKind["TypeAliasDeclaration"] = 266] = "TypeAliasDeclaration";
    SyntaxKind[SyntaxKind["EnumDeclaration"] = 267] = "EnumDeclaration";
    SyntaxKind[SyntaxKind["ModuleDeclaration"] = 268] = "ModuleDeclaration";
    SyntaxKind[SyntaxKind["ModuleBlock"] = 269] = "ModuleBlock";
    SyntaxKind[SyntaxKind["CaseBlock"] = 270] = "CaseBlock";
    SyntaxKind[SyntaxKind["NamespaceExportDeclaration"] = 271] = "NamespaceExportDeclaration";
    SyntaxKind[SyntaxKind["ImportEqualsDeclaration"] = 272] = "ImportEqualsDeclaration";
    SyntaxKind[SyntaxKind["ImportDeclaration"] = 273] = "ImportDeclaration";
    SyntaxKind[SyntaxKind["ImportClause"] = 274] = "ImportClause";
    SyntaxKind[SyntaxKind["NamespaceImport"] = 275] = "NamespaceImport";
    SyntaxKind[SyntaxKind["NamedImports"] = 276] = "NamedImports";
    SyntaxKind[SyntaxKind["ImportSpecifier"] = 277] = "ImportSpecifier";
    SyntaxKind[SyntaxKind["ExportAssignment"] = 278] = "ExportAssignment";
    SyntaxKind[SyntaxKind["ExportDeclaration"] = 279] = "ExportDeclaration";
    SyntaxKind[SyntaxKind["NamedExports"] = 280] = "NamedExports";
    SyntaxKind[SyntaxKind["NamespaceExport"] = 281] = "NamespaceExport";
    SyntaxKind[SyntaxKind["ExportSpecifier"] = 282] = "ExportSpecifier";
    SyntaxKind[SyntaxKind["MissingDeclaration"] = 283] = "MissingDeclaration";
    SyntaxKind[SyntaxKind["ExternalModuleReference"] = 284] = "ExternalModuleReference";
    SyntaxKind[SyntaxKind["JsxElement"] = 285] = "JsxElement";
    SyntaxKind[SyntaxKind["JsxSelfClosingElement"] = 286] = "JsxSelfClosingElement";
    SyntaxKind[SyntaxKind["JsxOpeningElement"] = 287] = "JsxOpeningElement";
    SyntaxKind[SyntaxKind["JsxClosingElement"] = 288] = "JsxClosingElement";
    SyntaxKind[SyntaxKind["JsxFragment"] = 289] = "JsxFragment";
    SyntaxKind[SyntaxKind["JsxOpeningFragment"] = 290] = "JsxOpeningFragment";
    SyntaxKind[SyntaxKind["JsxClosingFragment"] = 291] = "JsxClosingFragment";
    SyntaxKind[SyntaxKind["JsxAttribute"] = 292] = "JsxAttribute";
    SyntaxKind[SyntaxKind["JsxAttributes"] = 293] = "JsxAttributes";
    SyntaxKind[SyntaxKind["JsxSpreadAttribute"] = 294] = "JsxSpreadAttribute";
    SyntaxKind[SyntaxKind["JsxExpression"] = 295] = "JsxExpression";
    SyntaxKind[SyntaxKind["JsxNamespacedName"] = 296] = "JsxNamespacedName";
    SyntaxKind[SyntaxKind["CaseClause"] = 297] = "CaseClause";
    SyntaxKind[SyntaxKind["DefaultClause"] = 298] = "DefaultClause";
    SyntaxKind[SyntaxKind["HeritageClause"] = 299] = "HeritageClause";
    SyntaxKind[SyntaxKind["CatchClause"] = 300] = "CatchClause";
    SyntaxKind[SyntaxKind["ImportAttributes"] = 301] = "ImportAttributes";
    SyntaxKind[SyntaxKind["ImportAttribute"] = 302] = "ImportAttribute";
    SyntaxKind[SyntaxKind["PropertyAssignment"] = 303] = "PropertyAssignment";
    SyntaxKind[SyntaxKind["ShorthandPropertyAssignment"] = 304] = "ShorthandPropertyAssignment";
    SyntaxKind[SyntaxKind["SpreadAssignment"] = 305] = "SpreadAssignment";
    SyntaxKind[SyntaxKind["EnumMember"] = 306] = "EnumMember";
    SyntaxKind[SyntaxKind["SourceFile"] = 307] = "SourceFile";
    SyntaxKind[SyntaxKind["JSDocTypeExpression"] = 308] = "JSDocTypeExpression";
    SyntaxKind[SyntaxKind["JSDocNameReference"] = 309] = "JSDocNameReference";
    SyntaxKind[SyntaxKind["JSDocAllType"] = 310] = "JSDocAllType";
    SyntaxKind[SyntaxKind["JSDocNullableType"] = 311] = "JSDocNullableType";
    SyntaxKind[SyntaxKind["JSDocNonNullableType"] = 312] = "JSDocNonNullableType";
    SyntaxKind[SyntaxKind["JSDocOptionalType"] = 313] = "JSDocOptionalType";
    SyntaxKind[SyntaxKind["JSDocVariadicType"] = 314] = "JSDocVariadicType";
    SyntaxKind[SyntaxKind["JSDoc"] = 315] = "JSDoc";
    SyntaxKind[SyntaxKind["JSDocText"] = 316] = "JSDocText";
    SyntaxKind[SyntaxKind["JSDocTypeLiteral"] = 317] = "JSDocTypeLiteral";
    SyntaxKind[SyntaxKind["JSDocSignature"] = 318] = "JSDocSignature";
    SyntaxKind[SyntaxKind["JSDocLink"] = 319] = "JSDocLink";
    SyntaxKind[SyntaxKind["JSDocLinkCode"] = 320] = "JSDocLinkCode";
    SyntaxKind[SyntaxKind["JSDocLinkPlain"] = 321] = "JSDocLinkPlain";
    SyntaxKind[SyntaxKind["JSDocUnknownTag"] = 322] = "JSDocUnknownTag";
    SyntaxKind[SyntaxKind["JSDocAugmentsTag"] = 323] = "JSDocAugmentsTag";
    SyntaxKind[SyntaxKind["JSDocImplementsTag"] = 324] = "JSDocImplementsTag";
    SyntaxKind[SyntaxKind["JSDocDeprecatedTag"] = 325] = "JSDocDeprecatedTag";
    SyntaxKind[SyntaxKind["JSDocPublicTag"] = 326] = "JSDocPublicTag";
    SyntaxKind[SyntaxKind["JSDocPrivateTag"] = 327] = "JSDocPrivateTag";
    SyntaxKind[SyntaxKind["JSDocProtectedTag"] = 328] = "JSDocProtectedTag";
    SyntaxKind[SyntaxKind["JSDocReadonlyTag"] = 329] = "JSDocReadonlyTag";
    SyntaxKind[SyntaxKind["JSDocOverrideTag"] = 330] = "JSDocOverrideTag";
    SyntaxKind[SyntaxKind["JSDocCallbackTag"] = 331] = "JSDocCallbackTag";
    SyntaxKind[SyntaxKind["JSDocOverloadTag"] = 332] = "JSDocOverloadTag";
    SyntaxKind[SyntaxKind["JSDocParameterTag"] = 333] = "JSDocParameterTag";
    SyntaxKind[SyntaxKind["JSDocReturnTag"] = 334] = "JSDocReturnTag";
    SyntaxKind[SyntaxKind["JSDocThisTag"] = 335] = "JSDocThisTag";
    SyntaxKind[SyntaxKind["JSDocTypeTag"] = 336] = "JSDocTypeTag";
    SyntaxKind[SyntaxKind["JSDocTemplateTag"] = 337] = "JSDocTemplateTag";
    SyntaxKind[SyntaxKind["JSDocTypedefTag"] = 338] = "JSDocTypedefTag";
    SyntaxKind[SyntaxKind["JSDocSeeTag"] = 339] = "JSDocSeeTag";
    SyntaxKind[SyntaxKind["JSDocPropertyTag"] = 340] = "JSDocPropertyTag";
    SyntaxKind[SyntaxKind["JSDocThrowsTag"] = 341] = "JSDocThrowsTag";
    SyntaxKind[SyntaxKind["JSDocSatisfiesTag"] = 342] = "JSDocSatisfiesTag";
    SyntaxKind[SyntaxKind["JSDocImportTag"] = 343] = "JSDocImportTag";
    SyntaxKind[SyntaxKind["SyntaxList"] = 344] = "SyntaxList";
    SyntaxKind[SyntaxKind["JSTypeAliasDeclaration"] = 345] = "JSTypeAliasDeclaration";
    SyntaxKind[SyntaxKind["JSImportDeclaration"] = 346] = "JSImportDeclaration";
    SyntaxKind[SyntaxKind["NotEmittedStatement"] = 347] = "NotEmittedStatement";
    SyntaxKind[SyntaxKind["PartiallyEmittedExpression"] = 348] = "PartiallyEmittedExpression";
    SyntaxKind[SyntaxKind["SyntheticReferenceExpression"] = 349] = "SyntheticReferenceExpression";
    SyntaxKind[SyntaxKind["NotEmittedTypeElement"] = 350] = "NotEmittedTypeElement";
    SyntaxKind[SyntaxKind["Count"] = 351] = "Count";
    SyntaxKind[SyntaxKind["FirstAssignment"] = 63] = "FirstAssignment";
    SyntaxKind[SyntaxKind["LastAssignment"] = 78] = "LastAssignment";
    SyntaxKind[SyntaxKind["FirstCompoundAssignment"] = 64] = "FirstCompoundAssignment";
    SyntaxKind[SyntaxKind["LastCompoundAssignment"] = 78] = "LastCompoundAssignment";
    SyntaxKind[SyntaxKind["FirstReservedWord"] = 82] = "FirstReservedWord";
    SyntaxKind[SyntaxKind["LastReservedWord"] = 117] = "LastReservedWord";
    SyntaxKind[SyntaxKind["FirstKeyword"] = 82] = "FirstKeyword";
    SyntaxKind[SyntaxKind["LastKeyword"] = 166] = "LastKeyword";
    SyntaxKind[SyntaxKind["FirstFutureReservedWord"] = 118] = "FirstFutureReservedWord";
    SyntaxKind[SyntaxKind["LastFutureReservedWord"] = 126] = "LastFutureReservedWord";
    SyntaxKind[SyntaxKind["FirstTypeNode"] = 183] = "FirstTypeNode";
    SyntaxKind[SyntaxKind["LastTypeNode"] = 206] = "LastTypeNode";
    SyntaxKind[SyntaxKind["FirstPunctuation"] = 18] = "FirstPunctuation";
    SyntaxKind[SyntaxKind["LastPunctuation"] = 78] = "LastPunctuation";
    SyntaxKind[SyntaxKind["FirstToken"] = 0] = "FirstToken";
    SyntaxKind[SyntaxKind["LastToken"] = 166] = "LastToken";
    SyntaxKind[SyntaxKind["FirstLiteralToken"] = 8] = "FirstLiteralToken";
    SyntaxKind[SyntaxKind["LastLiteralToken"] = 14] = "LastLiteralToken";
    SyntaxKind[SyntaxKind["FirstTemplateToken"] = 14] = "FirstTemplateToken";
    SyntaxKind[SyntaxKind["LastTemplateToken"] = 17] = "LastTemplateToken";
    SyntaxKind[SyntaxKind["FirstBinaryOperator"] = 29] = "FirstBinaryOperator";
    SyntaxKind[SyntaxKind["LastBinaryOperator"] = 78] = "LastBinaryOperator";
    SyntaxKind[SyntaxKind["FirstStatement"] = 244] = "FirstStatement";
    SyntaxKind[SyntaxKind["LastStatement"] = 260] = "LastStatement";
    SyntaxKind[SyntaxKind["FirstNode"] = 167] = "FirstNode";
    SyntaxKind[SyntaxKind["FirstJSDocNode"] = 308] = "FirstJSDocNode";
    SyntaxKind[SyntaxKind["LastJSDocNode"] = 343] = "LastJSDocNode";
    SyntaxKind[SyntaxKind["FirstJSDocTagNode"] = 322] = "FirstJSDocTagNode";
    SyntaxKind[SyntaxKind["LastJSDocTagNode"] = 343] = "LastJSDocTagNode";
    SyntaxKind[SyntaxKind["FirstContextualKeyword"] = 127] = "FirstContextualKeyword";
    SyntaxKind[SyntaxKind["LastContextualKeyword"] = 166] = "LastContextualKeyword";
    SyntaxKind[SyntaxKind["LastUnaryOperator"] = 54] = "LastUnaryOperator";
    SyntaxKind[SyntaxKind["FirstTriviaToken"] = 2] = "FirstTriviaToken";
    SyntaxKind[SyntaxKind["LastTriviaToken"] = 6] = "LastTriviaToken";
})(SyntaxKind || (SyntaxKind = {}));
